lib/utils/agentconn/agent_windows.go

```diff
@@ -20,20 +20,228 @@ limitations under the License.
+			}
+			uid = 0x1000*x + rid
+		} else if len(sidParts) == 8 {
+			// SIDs from the local machine's SAM, the machine's primary
+			// domain, or a trusted domain of the machine's primary domain
+			uid, err = strconv.ParseInt(sidParts[7], 10, 32)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			unsureOfUID = true
+		} else {
+			// the format of the SID isn't supported, fallback to getting
+			// the UID from 'ps'
+			uid, err = getCygwinUIDFromPS()
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		}
+	}
+
+	// dial socket and complete handshake
+	var conn net.Conn
+	if !unsureOfUID {
+		// we're confident in what the Cygwin UID is, only make one attempt
+		// at establishing a connection
+		conn, err = attemptCygwinHandshake(port, key, uid)
+		if err == nil {
+			return conn, nil
+		}
+	} else {
+		// the Cygwin UID could be built a few different ways; attempt
+		// with all UIDs until one succeeds
+		cygwinRIDNums := []int64{0x30000, 0x100000, 0x80000000}
+		for _, num := range cygwinRIDNums {
+			conn, err = attemptCygwinHandshake(port, key, num+uid)
+			if err == nil {
+				return conn, nil
+			}
+		}
+
+		// none of those UIDs worked, fallback to getting UID from 'ps'
+		uid, err = getCygwinUIDFromPS()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		conn, err = attemptCygwinHandshake(port, key, uid)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	return conn, nil
+}
+
+// use Cygwin 'ps' binary to get the Cygwin UID of the current user
+func getCygwinUIDFromPS() (int64, error) {
+	// Cygwin 'bash' is used to call 'ps' so a Cygwin environment can be
+	// inherited by 'ps'
+	psOutput, err := exec.Command("bash.exe", "-c", "ps").Output()
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	psMatches := psLine.FindStringSubmatch(string(psOutput))
+	if len(psMatches) != 2 {
+		return 0, trace.Errorf("error reading Cygwin ps output")
+	}
+	uid, err := strconv.ParseInt(psMatches[1], 10, 32)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	return uid, nil
+}
+
+// connect to a listening socket of a Cygwin SSH agent and attempt to
+// preform a successful handshake with it. Handshake details here:
+// https://stackoverflow.com/questions/23086038/what-mechanism-is-used-by-msys-cygwin-to-emulate-unix-domain-sockets
+func attemptCygwinHandshake(port, key string, uid int64) (net.Conn, error) {
+	logrus.Debugf("[KEY AGENT] attempting a handshake with Cygwin ssh-agent socket; port=%s uid=%d", port, uid)
+
+	conn, err := net.Dial("tcp", "localhost:"+port)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// 1. send hex-encoded GUID
+	keyBuf := make([]byte, 16)
+	fmt.Sscanf(
+		key,
+		"%02x%02x%02x%02x-%02x%02x%02x%02x-%02x%02x%02x%02x-%02x%02x%02x%02x",
+		&keyBuf[3], &keyBuf[2], &keyBuf[1], &keyBuf[0],
+		&keyBuf[7], &keyBuf[6], &keyBuf[5], &keyBuf[4],
+		&keyBuf[11], &keyBuf[10], &keyBuf[9], &keyBuf[8],
+		&keyBuf[15], &keyBuf[14], &keyBuf[13], &keyBuf[12],
+	)
```

Comment:
I want to suggest something else here but I can't think of a better way to do this. Is the order of the  bytes here guaranteed?

Reply:
I thought the same thing when I first saw that, but I'm pretty sure the strange byte order here is guaranteed.

Reply:
Isn't this just little endian? I suspect a combination of `binary.LittleEndian` and `encoding/hex` would be easier to follow and more efficient than this massive string.

Reply:
Oh also, `go-winio` which we're already using has a `guid` package - I would check to see if there's something in there that already works.

Reply:
Yeah I suppose it is little endian, was initially thrown off at how the parts of the GUIDs were in the same order but the contents of each part were not. This GUID format is different than those in Windows so using something from `go-winio` or `golang.org/x/sys/windows` won't work unfortunately.